async_simple/util/ThreadPool.h

```diff
@@ -26,325 +26,130 @@
+                            [](auto &item) { return item.autoSchedule; }))
+                        break;
+                }
+            }
 
-inline ThreadPool::ThreadPool(size_t threadNum)
-    : _threadNum(threadNum),
-      _queue(threadNum ? threadNum : DEFAULT_THREADNUM),
-      _cond(threadNum ? threadNum : DEFAULT_THREADNUM),
-      _push(true),
-      _run(false),
-      _activeThreadCount(0) {
-    if (_threadNum == 0) {
-        _threadNum = DEFAULT_THREADNUM;
-    }
-}
+            if (!workerItem.fn && !_queues[id].pop(workerItem)) {
+                break;
+            }
 
-inline ThreadPool::~ThreadPool() { stop(STOP_AND_CLEAR_QUEUE); }
+            if (workerItem.fn) {
+                workerItem.fn();
+            }
+        }
+    };
 
-inline size_t ThreadPool::getItemCount() const {
-    size_t ret = 0;
-    for (size_t i = 0; i < _threadNum; ++i) {
-        ScopedLock lock(_cond[i]);
-        ret += _queue[i].size();
+    _threads.reserve(_threadNum);
+    for (auto i = 0; i < _threadNum; ++i) {
+        _threads.emplace_back(worker, i);
     }
-    return ret;
+}
+
+inline ThreadPool::~ThreadPool() {
+    _stop = true;
+    for (auto &queue : _queues)
+        queue.stop();
+    for (auto &thread : _threads)
+        thread.join();
 }
 
 inline ThreadPool::ERROR_TYPE ThreadPool::scheduleById(std::function<void()> fn,
                                                        int32_t id) {
-    if (-1 == id) {
-        id = rand() % _threadNum;
-    }
-    assert(id >= 0 && static_cast<size_t>(id) < _threadNum);
-    WorkItem *item = new WorkItem(std::move(fn));
-    if (!_push) {
-        return ERROR_POOL_HAS_STOP;
-    }
-
-    if (NULL == item) {
+    if (nullptr == fn) {
         return ERROR_POOL_ITEM_IS_NULL;
     }
 
-    ScopedLock lock(_cond[id]);
-
-    if (!_push) {
+    if (_stop) {
         return ERROR_POOL_HAS_STOP;
     }
 
-    _queue[id].push(item);
-    _cond[id].signalConsumer();
-    return ERROR_NONE;
-}
-
-inline bool ThreadPool::start() {
-    if (_run) {
-        return false;
-    }
-
-    _push = true;
-    _run = true;
+    if (id == -1) {
+        if (_enableWorkSteal) {
+            // Try to push to a non-block queue firstly.
+            WorkItem workerItem{false, fn};
+            for (auto n = 0; n < _threadNum * 2; ++n) {
+                if (_queues.at(n % _threadNum).try_push(workerItem))
+                    return ERROR_NONE;
+            }
+        }
 
-    if (createThreads()) {
-        return true;
+        id = rand() % _threadNum;
+        _queues[id].push(WorkItem{false, std::move(fn)});
     } else {
-        stop(STOP_THREAD_ONLY);
-        return false;
-    }
-}
-
-inline void ThreadPool::stop(STOP_TYPE stopType) {
-    if (STOP_THREAD_ONLY != stopType) {
-        { _push = false; }
-    }
-
-    if (STOP_AFTER_QUEUE_EMPTY == stopType) {
-        waitQueueEmpty();
+        assert(id < _threadNum);
+        _queues[id].push(WorkItem{true, std::move(fn)});
```

Comment:
WorkItem{true means this task can be steal by other thread?

Reply:
Do a improvement:
```
    if (id == -1) {
        if (_enableWorkSteal) {
            // Try to push to a non-block queue firstly.
            WorkItem workerItem{/*canSteal = */ true, fn};
            for (auto n = 0; n < _threadNum * 2; ++n) {
                if (_queues.at(n % _threadNum).try_push(workerItem))
                    return ERROR_NONE;
            }
        }

        id = rand() % _threadNum;
        _queues[id].push(
            WorkItem{/*canSteal = */ _enableWorkSteal, std::move(fn)});
    } else {
        assert(id < _threadNum);
        _queues[id].push(WorkItem{/*canSteal = */ false, std::move(fn)});
    }
```
id is -1 and _enableWorkSteal true: can worksteal;
otherwise: never worksteal;

Is this code better and clear?

Reply:
👍🏻  nice